### PR TITLE
dungeons and pv bingo tab

### PR DESCRIPTION
**Pull request by Lulonaut**
Some broken dungeon things fixed by nopo and jani

bingo tab in pv and movable pv tabs

### CODE REVIEW

**Review by NopoTheGamer — APPROVED** [review]
https://www.google.com/url?sa=i&url=https%3A%2F%2Fwww.shutterstock.com%2Fsearch%2Fbingo%2Bcard&psig=AOvVaw3A5QHJMMkTCT3v-ws0DfCS&ust=1647171123861000&source=images&cd=vfe&ved=0CAsQjRxqFwoTCIjJ1eS8wPYCFQAAAAAdAAAAABAD

**Comment by Lulonaut** [comment]
Thank you nopo

**Review by NopoTheGamer — APPROVED** [review]
Thank you hypixel wiki for just being wrong

**Review by NopoTheGamer — DISMISSED** [review]

**Review by DeDiamondPro — DISMISSED** [review]

**Review by DeDiamondPro — DISMISSED** [review]

**Review by NopoTheGamer — APPROVED** [review]
"define productive" math homework

**Comment by Justin2528** [comment]
ok @NopoTheGamer

File: Update Notes/2.1.md [hunk 1]
```diff
@@ -13,9 +13,10 @@
 - [Added an armor overlay for the new armor slots](https://cdn.discordapp.com/attachments/832652653292027904/922399046528794634/unknown.png) 
 - Added a pet overlay that shows your active pet in your inventory
 - [Price graph for items on /ah and /bz](https://cdn.discordapp.com/attachments/896407218151366687/926968296929107999/unknown.png) - DeDiamondPro
+- [Added bingo tab in /pv](https://cdn.discordapp.com/attachments/702456294874808330/951909539383500820/unknown.png) - Lulonaut
 - Added wishing compass solver that shows target coordinates using two samples - CraftyOldMiner
 - Improved metal detector logic to solve using a single position in most cases using known locations based on Keeper coordinates - CraftyOldMiner
-### **Minor Changes:**
```
Comment (romangraef) [inline Update Notes/2.1.md:18]: :( i liked that colon

File: src/main/java/io/github/moulberry/notenoughupdates/core/config/gui/GuiOptionEditorDraggableList.java [hunk 2]
```diff
@@ -35,9 +36,14 @@
...
 
-	public GuiOptionEditorDraggableList(ConfigProcessor.ProcessedOption option, String[] exampleText) {
+	public GuiOptionEditorDraggableList(
+		ConfigProcessor.ProcessedOption option,
+		String[] exampleText,
+		boolean disableDeleting
+	) {
 		super(option);
 
+		this.enableDeleting = disableDeleting;
```
Comment (romangraef) [inline src/main/java/io/github/moulberry/notenoughupdates/core/config/gui/GuiOptionEditorDraggableList.java:46]: This is weird naming of the param.

File: src/main/java/io/github/moulberry/notenoughupdates/miscfeatures/AuctionBINWarning.java [hunk 3]
```diff
@@ -105,13 +109,16 @@ public boolean onMouseClick(Slot slotIn, int slotId, int clickedButton, int clic
 				lowestPrice = (int) NotEnoughUpdates.INSTANCE.manager.auctionManager.getItemAvgBin(internalname);
 			}
 
-			//TODO: Add option for warning if lowest price does not exist
+			float undercutFactor = 1 - NotEnoughUpdates.INSTANCE.config.ahTweaks.warningThreshold / 100;
+			if (undercutFactor < 0) undercutFactor = 0;
+			if (undercutFactor > 1) undercutFactor = 1;
+			float overcutFactor = 1 - NotEnoughUpdates.INSTANCE.config.ahTweaks.overcutWarningThreshold / 100;
```
Comment (romangraef) [inline src/main/java/io/github/moulberry/notenoughupdates/miscfeatures/AuctionBINWarning.java:115]: Perhaps this should be `1 + ...` with the coercing adjusted accordingly? Not tested this out, but just looking at the config option description this feels wrong.

File: src/main/java/io/github/moulberry/notenoughupdates/options/seperateSections/Misc.java [hunk 4]
```diff
@@ -107,6 +113,7 @@
 
 	@Expose
 	@ConfigOption(
+
```
Comment (romangraef) [inline src/main/java/io/github/moulberry/notenoughupdates/options/seperateSections/Misc.java:116]: nice whitespace

File: src/main/java/io/github/moulberry/notenoughupdates/profileviewer/BingoPage.java [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+
+	private static void processBingoResources() {
+		long currentTime = System.currentTimeMillis();
+
+		//renew every 2 minutes
+		if (currentTime - lastResourceRequest < 120 * 1000 && bingoGoals != null) return;
+		lastResourceRequest = currentTime;
+
+		HashMap<String, String> args = new HashMap<>();
+		NotEnoughUpdates.INSTANCE.manager.hypixelApi.getHypixelApiAsync(
```
Comment (romangraef) [inline src/main/java/io/github/moulberry/notenoughupdates/profileviewer/BingoPage.java:298]: consider using `getHypixelApiAsync(String, String, HashMap<String, String>)` (the `CompletableFuture`), and checking for `CompletableFuture#isDone()`. this approach is also okay if it has to be this way.

File: src/main/java/io/github/moulberry/notenoughupdates/profileviewer/GuiProfileViewer.java [hunk 6]
```diff
@@ -1688,150 +1963,35 @@ private void renderXpBar(
...
-		}
-		PetLevel levelObj = new PetLevel();
-		levelObj.level = level;
-		levelObj.currentLevelRequirement = currentLevelRequirement;
-		levelObj.maxXP = xpTotal;
-		levelObj.levelPercentage = currentLevelProgress / currentLevelRequirement;
-		levelObj.levelXp = currentLevelProgress;
-		levelObj.totalXp = exp;
-		return levelObj;
+	private ItemStack getQuestionmarkSkull() {
```
Comment (romangraef) [inline src/main/java/io/github/moulberry/notenoughupdates/profileviewer/GuiProfileViewer.java:1966]: Im pretty sure this is just rearranging code, but we also have `Utils#createSkull` for this now.

**Review by romangraef — CHANGES_REQUESTED** [review]

File: Update Notes/2.1.md (hunk 1 above)
Comment (Lulonaut) [reply Update Notes/2.1.md:18]: thats the only colon in the entire file :slightly_smiling_face:

Comment (romangraef) [reply Update Notes/2.1.md:18]: soon i will spread your colon across that file as well if you dont undo that deletion :imp:

File: src/main/java/io/github/moulberry/notenoughupdates/miscfeatures/AuctionBINWarning.java (hunk 3 above)
Comment (romangraef) [reply src/main/java/io/github/moulberry/notenoughupdates/miscfeatures/AuctionBINWarning.java:115]: oh while im at it: there isnt really a reason to limit the overcut percentage to 100% max, is there?

**Review by DeDiamondPro — DISMISSED** [review]

**Review by romangraef — APPROVED** [review]

**Review by NopoTheGamer — APPROVED** [review]

**Review by DeDiamondPro — APPROVED** [review]

**Pull request merged**